rfc/drafts/indexer-ideas-v1.md

```diff
@@ -0,0 +1,303 @@
+  - E.g. `EntityLinkCell = {"group_entity_link":[(ref1,ref2,ref3),(ref4,ref5),(ref6)]}`
+  - Meaning that all grouped entities in that list should have the indicated relationship, and all separated
+    entities should have the inverse relationship. I.e. is_derivation vs is_not_derivation.
+  - Could be easily converted to all-pairs of links internally.
+
++ Object-localization for video / audio / images?
+  - System identifies bounding-boxes for multiple objects in an image / time ranges for multiple known audio clips
+    in an audio broadcast / time ranges for known video clips.
+  - System allows user to focus on just one of those detected objects in their search / dedupe query.
+  - Relevant to the search / dedupe architecture, because this likely would require for those multiple objects
+    to be indexed separately, along with their location in the media.
+
++ Non-monetary incentives for resource providers? I.e. User donates resources because he likes particular content.
+  - Force resource providers to support all content equally (Freenet).
+  - Allow full control over the content you're supporting (Bittorrent).
+  - Something in the middle (Tor exit nodes)?
+
++ Resource usage considerations?
+  - In non-P2P systems, the equation to optimize for tends to be simple. E.g. maximizing utilization of already-owned hardware,
+    or maximizing per-dollar results.
+  - Costs can be different in P2P systems. Some users can get upset and leave, if all of their resources are
+    immediately maxed out.
+  - Considering that P2P networks tend to have a power-law distribution of resources.
+
++ Allowing user to explicitly describe the type of semantic-similarity he's searching for?
+  - Query interpretation was part of textual search engines since very early on - "do know go", but this isn't done for the major image search engines?
+  - Allow users to directly tell the system what kind of semantic similarity they're searching for?
+    - Some indexing systems don't natively deal with multiple types of semantic similarity, or worse, blur multiple types together into an average that doesn't resemble any meaningful type of semantic similarity.
+    - One type of interface: user inputs image - enters a prose description of the type of similarity he's after?
+      - pictures containing same brands and trademarks as this picture
+      - more pictures in the style of this artist
+      - people with same pose as this picture
+      - same type of coffee mug as this picture
+      - more pictures of from this location
+      - more pictures of this person
+      - mashups containing this picture
+  - Or, we start with a small number of specific types of semantic similarity to start.
+  - Range of ways to support this, some relatively simple.
+
++ Eviction of images from the index?
+
++ Co-location of images that should be stored together, for faster retrieval / bulk deletion / cache promotion or eviction?
+
+
+## V1 SEARCH / DEDUPE API:
```

Comment:
Yes, seems to me that `ArtefactLinkCell` is a missing piece that we may need.

Perhaps we should create either a table / list of pseudo-code function definitions, for all the graph edge / node operations, along with their currently-known required and optional arguments?

A few questions I'm have after looking at the code / RFC1 - 

```
  - entity $roundTripEntity
  - artefact $roundTripArtefact
  - entity chain cell $roundTripEntityChainCell
  - entity update cell $roundTripEntityUpdateCell
  - entity link cell $roundTripEntityLinkCell
  - artefact chain cell $roundTripArtefactChainCell
  - artefact update cell $roundTripArtefactUpdateCell
  - artefact creation cell $roundTripArtefactCreationCell
  - artefact derivation cell $roundTripArtefactDerivationCell
  - artefact ownership cell $roundTripArtefactOwnershipCell
  - artefact reference cell $roundTripArtefactReferenceCell
  - canonical journal entry $roundTripCanonicalEntry
  - chain journal entry $roundTripChainEntry
  - journal block $roundTripJournalBlock
```

Interpreting those:

  - ArtefactCreation - create entity (what about artefact?) node.
  - EntityUpdate / ArtefactUpdate - update entity or artefact node.
  - EntityLink - create edge between (entity <-> entity), as specified by "relationship".
  - ArtefactOwnership - create undirected edge between (entity <-> artefact).
  - ArtefactDerivation - create directed edge from (artefact --> artefact).
  - ArtefactReference - create undirected edge from (artefact <-> external_identifier).

So, questions:

  - There's ArtefactCreation but no EntityCreation?
  - EntityLink allows you to create graph edges between 2 entities, with an arbitrarily specified "relationship" string, but other edge types (ArtefactOwnership, ArtefactDerivation, ArtefactReference) are defined in a different way?
  - Undirected (artefact <-> artefact) links? E.g. dedupe artefacts, e.g. ArtefactLinkCell?
  - Undirected (entity <-> entity) links? E.g. dedupe entites?
      + Are 2 different types of of artist dedupe needed? One where the exact same persona is present twice, and one where an artist is using multiple personas? 
  -  Are entities intended only for artists, or to also represent users in the system (who will never have artefact ownership associated with them)?
  - Can edges (ArtefactOwnership, ArtefactDerivation, ArtefactReference, EntityLink) receive updates, or only nodes (entities, artefacts)? Does the user just create an object again at a later point in the journal in order to update it?